### PR TITLE
Update consul statefulset

### PR DESCRIPTION
**Pull request by lachie83**
This PR does the following:

- Bumps Consul version to 0.7.2
- Add conditional for enabling the UI and an optional Kubernetes service for access to the UI
- Migrates from PetSet to StatefulSet
- Adds helper functions and NOTES.txt
- Lowers the memory request
- Removes unneeded annotations referencing helm.sh
- Removed readinessprobe and replaced with liveliness probe
- Update gotemplates to support longer names
- Moved to stable

### CODE REVIEW

File: stable/consul/templates/consul.yaml [hunk 1]
```diff
@@ -157,17 +182,11 @@ spec:
...
-          path: /v1/health/service/consul
-          port: {{.Values.HttpPort}}
-        initialDelaySeconds: 10
-        timeoutSeconds: 1
   volumeClaimTemplates:
   - metadata:
       name: datadir
       annotations:
-        volume.alpha.kubernetes.io/storage-class: anything
+        volume.beta.kubernetes.io/storage-class: "{{.Values.StorageClass}}"
```
Comment (viglesiasce) [inline stable/consul/templates/consul.yaml]: Can we make this logic the same as in this PR #367? (ie alpha if not provided)

**Review by viglesiasce — CHANGES_REQUESTED** [review]

File: stable/consul/templates/consul.yaml (hunk 1 above)
Comment (lachie83) [reply stable/consul/templates/consul.yaml]: I've completed this although I'm not sure if it should be here. Reasoning is that StatefulSet is 1.5 and the corresponding storage annotation is now beta. 1.4 used PetSet and it used the storage used alpha notation. Thoughts?

Comment (viglesiasce) [reply stable/consul/templates/consul.yaml]: The change is more about the default storage class than alpha vs beta features. With the alpha you don't need a default cluster level storage class while with the beta you do. Until we have more clusters that have default storage classes I think we should use the alpha+beta pattern.

Comment (lachie83) [reply stable/consul/templates/consul.yaml]: Okay. Appreciate the explanation. This change should meet that requirement.

**Comment by stevesloka** [comment]
When I deploy I get the following error now after creating my storage-class by hand manually: 

```
Waiting for consul3-consul-1.consul3-consul to come up
ping: bad address 'consul3-consul-1.consul3-consul.default.svc.cluster.local'
```

**Comment by lachie83** [comment]
@stevesloka can you confirm if the first pod (consul-0) is in state running 1/1? Could you also grab the k8s events on the consul-0 pod.

**Comment by stevesloka** [comment]
My consul-0 pod when running then crashed: 

```
NAME               READY     STATUS             RESTARTS   AGE
consul3-consul-0   0/1       CrashLoopBackOff   13         2h
consul3-consul-1   0/1       Pending            0          31m
```

Here are the events:
```
LASTSEEN   FIRSTSEEN   COUNT     NAME               KIND      SUBOBJECT                         TYPE      REASON             SOURCE                                  MESSAGE
35m        2h          406       consul3-consul-0   Pod                                         Warning   FailedScheduling   {default-scheduler }                    [SchedulerPredicates failed due to PersistentVolumeClaim is not bound: "datadir-consul3-consul-0", which is unexpected., SchedulerPredicates failed due to PersistentVolumeClaim is not bound: "datadir-consul3-consul-0", which is unexpected., SchedulerPredicates failed due to PersistentVolumeClaim is not bound: "datadir-consul3-consul-0", which is unexpected.]
34m        34m         1         consul3-consul-0   Pod                                         Normal    Scheduled          {default-scheduler }                    Successfully assigned consul3-consul-0 to ip-10-0-72-153.ec2.internal
4m         33m         14        consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Pulling            {kubelet ip-10-0-72-153.ec2.internal}   pulling image "consul:0.7.2"
4m         33m         14        consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Pulled             {kubelet ip-10-0-72-153.ec2.internal}   Successfully pulled image "consul:0.7.2"
33m        33m         1         consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Created            {kubelet ip-10-0-72-153.ec2.internal}   Created container with docker id b855e7d7dc66; Security:[seccomp=unconfined]
33m        33m         1         consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Started            {kubelet ip-10-0-72-153.ec2.internal}   Started container with docker id b855e7d7dc66
4m         33m         16        consul3-consul-0   Pod       spec.containers{consul3-consul}   Warning   Unhealthy          {kubelet ip-10-0-72-153.ec2.internal}   Liveness probe failed: Error connecting to Consul agent: dial tcp 127.0.0.1:8400: getsockopt: connection refused

32m       32m       1         consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Killing      {kubelet ip-10-0-72-153.ec2.internal}   Killing container with docker id b855e7d7dc66: pod "consul3-consul-0_default(5ca9ac46-d819-11e6-9cdb-0a8828a2b648)" container "consul3-consul" is unhealthy, it will be killed and re-created.
32m       32m       1         consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Created      {kubelet ip-10-0-72-153.ec2.internal}   Created container with docker id b8bd70e5a0a9; Security:[seccomp=unconfined]
32m       32m       1         consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Started      {kubelet ip-10-0-72-153.ec2.internal}   Started container with docker id b8bd70e5a0a9
31m       31m       1         consul3-consul-0   Pod       spec.containers{consul3-consul}   Normal    Killing      {kubelet ip-10-0-72-153.ec2.internal}   Killing container with docker id b8bd70e5a0a9: pod "consul3-consul-0_default(5ca9ac46-d819-11e6-9cdb-0a8828a2b648)" container "consul3-consul" is unhealthy, it will be killed and re-created.
```

**Comment by lachie83** [comment]
Thanks @stevesloka Okay the liveliness probe is killing the container. Can you test removing it from the manifest?

**Comment by stevesloka** [comment]
Yeah @lachie83 after removing that all works well. 

```
consul5-consul-0  10.2.90.2:8301   alive   server  0.7.2  2         dc1
consul5-consul-1  10.2.2.2:8301    alive   server  0.7.2  2         dc1
consul5-consul-2  10.2.101.7:8301  alive   server  0.7.2  2         dc1
```

Total side question, do you know of a way to enforce zone distribution (short of writing your own scheduler), to make sure 1 member goes into each AZ? This is especially important with the EBS volumes on AWS since we can't **easily** move them.

**Comment by lachie83** [comment]
@stevesloka -- Yes, do this -- https://github.com/kow3ns/charts/blob/47e2ab31506d0e788911c67c123798fca0f67ddf/incubator/zookeeper/templates/ss.yaml#L22. I can add this to the consul chart if it would be useful.

**Comment by lachie83** [comment]
@stevesloka any ideas on a better health check? All my ideas assume that the cluster is up.

**Comment by stevesloka** [comment]
@lachie83 Sorry haven't had a chance to get back to this. Is there a way to know if the member is healthy? I guess checking against the other members is that check which is why the issue exists.

**Comment by mgoodness** [comment]
Resolves #633

**Comment by lachie83** [comment]
Closing in favor of #671

**Pull request closed without merging**